rsync_to_rdisc.py

```diff
@@ -215,31 +214,48 @@ def rsync_server_remote(hpc_server, client, to_be_transferred, mount_path, run_f
     rsync_succes = True
 
     for run in to_be_transferred:
-        with open(settings.log_path, 'a', newline='\n') as log_file:
-            log_file_writer = writer(log_file, delimiter='\t')
+        with open(settings.log_path, "a", newline="\n") as log_file:
+            log_file_writer = writer(log_file, delimiter="\t")
             log_file_writer.writerows([["#########"], [f"Date: {date}"], [f"Run_folder: {run}"]])
 
         transfer_settings = to_be_transferred[run]
-        # settings per folder data type, such as remote input dir and local output dir, etc.
+        # Settings per folder data type, such as remote input dir and local output dir, etc.
         missing = check_if_file_missing(transfer_settings["files_required"], f"{transfer_settings['input']}/{run}", client)
 
         if missing:
             print(transfer_settings, rsync_succes, missing, run, run_file)
             rsync_succes = action_if_file_missing(transfer_settings, rsync_succes, missing, run, run_file)
             continue  # don't transfer the run if a required file is missing.
-        os.system((
-            "rsync -rahuL --stats {user}@{server}:{input}/{run} {mount_path}/{output}/ "
-            " 1>> {log_path} 2>> {errorlog_path} 2> {temp_error_path}").format(
-                user=settings.user,
-                server=hpc_server,
-                input=transfer_settings["input"],
-                run=run,
-                mount_path=mount_path,
-                output=transfer_settings["output"],
-                log_path=settings.log_path,
-                errorlog_path=settings.errorlog_path,
-                temp_error_path=settings.temp_error_path
-        ))
+
+        # Get include and exclude patterns as list to easily access key-value pair in unit test.
+        if transfer_settings.get("include", None):
+            include_patterns = [f"--include '{pattern}'" for pattern in transfer_settings["include"]]
+        else:
+            include_patterns = ""
+        if transfer_settings.get("exclude", None):
+            exclude_patterns = [f"--exclude '{pattern}'" for pattern in transfer_settings["exclude"]]
+        else:
+            exclude_patterns = ""
+
+        source_destination = f"{settings.user}@{hpc_server}:{transfer_settings['input']}/{run}"
+        target_path = f"{mount_path}/{transfer_settings['output']}"
+        parse_stdout = f"1>> {settings.log_path} 2>> {settings.errorlog_path} 2> {settings.temp_error_path}"
+
+        rsync_cmd = [
+            "rsync",
+            "-rahuL",
+            "--stats",
+            source_destination,
+            target_path,
+            parse_stdout
+        ]
+
+        # Using list slicing to extend rsync_cmd list with include / exclude patterns.
+        rsync_cmd[3:3] = exclude_patterns
+        # Add include patterns before exclude patterns.
+        rsync_cmd[3:3] = include_patterns
+
+        subprocess.run(rsync_cmd, shell=True, stdout=subprocess.PIPE, encoding="UTF-8")
```

Comment:
A small remark is that this variable is called `parse_stdout`, but it seems to do three different things.

1) it redirects (appends) the **standard output** to a log file
2) It redirects (appends) the **standard error** to a error log file
3) it redirects the **standard error** to a temporary path

None of them is technically 'parsing' anything, like the variable name suggests.

subprocess actually has built-in functions to handle output piping. 

```suggestion
        rsync_cmd = [
            "rsync",
            "-rahuL",
            "--stats",
            source_destination,
            target_path
        ]

        # Using list slicing to extend rsync_cmd list with include / exclude patterns.
        rsync_cmd[3:3] = exclude_patterns
        # Add include patterns before exclude patterns.
        rsync_cmd[3:3] = include_patterns

        with open(settings.log_path, 'w') as out, open(settings.errorlog_path, 'w') as err: # Or replace 'w' with 'a' when appending
            subprocess.run(rsync_cmd, shell=True, stdout=out, stderr=err, encoding="UTF-8")
```

Reply:
Another improvement is to validate the return code of the subprocess command. I didn't add this as a suggestion to avoid suggesting multiple things at once, but you could do:

```
result = subprocess.run(rsync_cmd, shell=True, stdout=out, stderr=err, encoding="UTF-8")
if result.returncode != 0:
    # throw an error or something
```